host/src/udf_query.rs

```diff
@@ -0,0 +1,168 @@
+//! Embedded SQL approach for executing Python UDFs within SQL queries.
+
+use datafusion::execution::TaskContext;
+use datafusion_common::{DataFusionError, Result as DataFusionResult};
+use datafusion_sql::parser::{DFParserBuilder, Statement};
+use sqlparser::ast::{CreateFunctionBody, Expr, Statement as SqlStatement, Value};
+use sqlparser::dialect::dialect_from_str;
+
+use crate::{WasmComponentPrecompiled, WasmScalarUdf};
+
+/// A SQL query containing a Python UDF and SQL string that uses the UDF
+#[derive(Debug, Clone)]
+pub struct UdfQuery(String);
+
+impl UdfQuery {
+    /// Create a new UDF query
+    pub fn new(query: String) -> Self {
+        Self(query)
+    }
+}
+
+impl AsRef<str> for UdfQuery {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+/// A [UdfQueryParseResult] contains the set of registered UDFs and the SQL query string.
+type UdfQueryParseResult = DataFusionResult<(Vec<WasmScalarUdf>, String)>;
+
+/// Handles the registration and invocation of UDF queries in DataFusion with a
+/// pre-compiled WASM component.
+pub struct UdfQueryParser<'a> {
+    /// Pre-compiled WASM component.
+    /// Necessary to create UDFs.
+    component: &'a WasmComponentPrecompiled,
```

Comment:
Could we make this one a language->component map, so this code is totally abstracted away from Python?


```suggestion
    components: HashMap<&'a str, &'a WasmComponentPrecompiled>,
```